### PR TITLE
Define room version characteristics and types

### PR DESCRIPTION
**Pull request by neilalexander**
This defines the core characteristics of each room version, and the exported types that Dendrite will use to implement them.

### CODE REVIEW

File: eventversion.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		stateResAlgorithm:      StateResV2,
+		eventFormat:            EventFormatV2,
+		eventIDFormat:          EventIDFormatV3,
+		enforceSignatureChecks: true,
+	},
+}
+
+// RoomVersionMeta contains information about a given room version, e.g. which
+// state resolution algorithm or event ID format to use.
+type roomVersion struct {
```
Comment (Kegsay) [inline eventversion.go]: This isn't called `RoomVersionMeta`. Pick a name, any name...

Comment (neilalexander) [reply eventversion.go]: Forgot to update the comment previously. It's fixed now.

File: eventversion.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	eventIDFormat          EventIDFormat
+	enforceSignatureChecks bool
+}
+
+// StateResAlgorithm returns the state resolution for the given room version.
+func (v RoomVersion) StateResAlgorithm() StateResAlgorithm {
+	if r, ok := roomVersionMeta[v]; ok {
+		return r.stateResAlgorithm
+	}
+	return StateResV1
```
Comment (Kegsay) [inline eventversion.go]: Why do we fall back to v1 if the version is bogus? Also, you can probably make this a bit cleaner if `roomVersionMeta[""]` had the defaults in it, then you just `return r.stateResAlgorithm`.

Comment (Kegsay) [reply eventversion.go]: Though really indexing with an invalid version should just panic.

Comment (neilalexander) [reply eventversion.go]: Made it panic instead if the version is unrecognised.

**Comment by Kegsay** [comment]
We *have* to check the room version is a version we know how to handle, otherwise fail gracefully. The current API shape doesn't make this easy, since it defaults to V1 values on bogus versions.

**Comment by neilalexander** [comment]
I suppose the question is whether we should `panic()` or return an `error` for each of those functions.

**Comment by neilalexander** [comment]
I've implemented errors now rather than panicking.

**Pull request merged**